src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultFileChangeTracker.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Editor.Razor;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    internal class DefaultFileChangeTracker : FileChangeTracker, IVsFileChangeEvents
+    {
+        private const uint FileChangeFlags = (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Del | _VSFILECHANGEFLAGS.VSFILECHG_Add);
+
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly ErrorReporter _errorReporter;
+        private readonly IVsFileChangeEx _fileChangeService;
+        private uint _fileChangeCookie;
+
+        public override event EventHandler<FileChangeEventArgs> Changed;
+
+        public DefaultFileChangeTracker(
+            string filePath,
+            ForegroundDispatcher foregroundDispatcher,
+            ErrorReporter errorReporter,
+            IVsFileChangeEx fileChangeService)
+        {
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(filePath));
+            }
+
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (errorReporter == null)
+            {
+                throw new ArgumentNullException(nameof(errorReporter));
+            }
+
+            if (fileChangeService == null)
+            {
+                throw new ArgumentNullException(nameof(fileChangeService));
+            }
+
+            FilePath = filePath;
+            _foregroundDispatcher = foregroundDispatcher;
+            _errorReporter = errorReporter;
+            _fileChangeService = fileChangeService;
+            _fileChangeCookie = VSConstants.VSCOOKIE_NIL;
+        }
+
+        public override string FilePath { get; }
+
+        public override void StartListening()
+        {
```

Comment:
Should assert foreground thread here or lock.

Reply:
👍.  Here and in `StopListening`